binary_search_tree/tree.py

```diff
@@ -14,44 +14,139 @@ class Tree:
+                return current.value
+            elif current.key > key:
+                current = current.left
+            else:
+                current = current.right
+        return None
+
+    def inorder_helper(self, current_node, items):
+        if current_node is None:
+            return []
+
+        self.inorder_helper(current_node.left, items)
+        items.append({"key": current_node.key, "value": current_node.value})
+        self.inorder_helper(current_node.right, items)
+
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def inorder(self):
-        pass
 
-    # Time Complexity: 
-    # Space Complexity:     
+        if self.root == None:
+            return []
+        
+        items = []
+        self.inorder_helper(self.root, items)
+        return items
+
+    def preorder_helper(self, current_node, items):
+        if current_node is None:
+            return []
+
+        items.append({"key": current_node.key, "value": current_node.value})
+        self.preorder_helper(current_node.left, items)
+        self.preorder_helper(current_node.right, items)
+
+    # Time Complexity: O(log n)
+    # Space Complexity: O(log n)
     def preorder(self):
-        pass
+        if self.root == None:
+            return []
+
+        items = []
+        self.preorder_helper(self.root, items)
+        return items
 
-    # Time Complexity: 
-    # Space Complexity:     
+    def postorder_helper(self, current_node, items):
+        if current_node is None:
+            return []
+
+        self.postorder_helper(current_node.left, items)
+        self.postorder_helper(current_node.right, items)
+        items.append({"key": current_node.key, "value": current_node.value})
+
+
+    # Time Complexity: O(log n)
+    # Space Complexity: O(log n) 
     def postorder(self):
-        pass
+        if self.root == None:
+            return []
+
+        items = []
+        self.postorder_helper(self.root, items)
+        return items
+
+
+    def height_helper(self, current_node):
+        if current_node is None:
+            return 0
+        left_height = self.height_helper(current_node.left)
+        right_height = self.height_helper(current_node.right)
+        return max(left_height, right_height) + 1
 
-    # Time Complexity: 
-    # Space Complexity:     
+
+    # Time Complexity: O(log n)
+    # Space Complexity: O(log n)  
     def height(self):
-        pass
+        if self.root == None:
+            return 0
+
+        return self.height_helper(self.root)
+
 
+    def bfs_helper(self, current_node, items):
+        if current_node is None:
+            return []
+
+        items.append({"key": current_node.key, "value": current_node.value})
+        self.bfs_helper(current_node.left, items)
+        self.bfs_helper(current_node.right, items)
 
 #   # Optional Method
-#   # Time Complexity: 
-#   # Space Complexity: 
+#   # Time Complexity: O(n)
+#   # Space Complexity: O(n)
     def bfs(self):
```

Comment:
With breadth first search you want to traverse the tree starting from its root, then visit each of it's children (from left to right), before visiting the children's children and so forth. In other words, you want to traverse the tree level by level. 

With your `bfs_helper`, you are adding the root node of the current tree, but then your recursive call through the left subtree will finish before you ever look at the root of the right subtree. 

(Hint: this may be easier to think about iteratively)